pkg/system/clock.go

```diff
@@ -0,0 +1,32 @@
+/*
+Copyright 2018 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package system
+
+import (
+	"time"
+)
+
+// Mockable interface for time based testing
```

Comment:
I'm not sure we need a mockable clock to test this, since we can just set `lastPinned` a ways in the past

Reply:
... or the future

Reply:
My hope was to get testing at the unit level similar to the e2e testing you mention above, but I had some difficulty wrapping my head around the reconciler table testing machinery. High level id really like to create some resources similar to the table tests, advance the clock manually, then assert a delete is called. Ideally I'd like to not actually 'run' the reconciler but set some state, call reconcile, assert, repeat. Curious if you have any thoughts on how to do that or if you'd really prefer to only hit this case in e2e tests for now.

Reply:
@greghaynes at least today each of the Reconcile's in the example you give would be a different row of the table test.  I know @dprotaso is trying to rework some of the table test stuff [here](https://github.com/knative/serving/pull/2074).

Reply:
Sorry, I may not be explaining this super clearly or maybe I'm missing something obvious :)

What happens is the tables (specifically this one https://github.com/knative/serving/pull/2000/files#diff-27d72d7a073dc66f0609e9c90e1e81ebR43) are generated in bulk before any tests are run, which is when we fill in WantPatches, etc with static values. As a result we have to set the lastPinned annoation to time.Now() at this point if were not mocking out a clock. Then when the tests run they take >1sec which results in most of them failing due to time.Now() at assertion time being different.

Reply:
I see, thanks for explaining.  That makes sense.